cmd/kubeadm/app/apis/kubeadm/validation/validation.go

```diff
@@ -54,61 +57,141 @@ func ValidateMasterConfiguration(c *kubeadm.MasterConfiguration) field.ErrorList
+	allErrs = append(allErrs, ValidateToken(c.Token, fldPath)...)
+	allErrs = append(allErrs, ValidateDiscoveryFile(c, fldPath)...)
+	allErrs = append(allErrs, ValidateToken(c.TLSBootstrapToken, fldPath)...)
+	allErrs = append(allErrs, ValidateJoinDiscoveryTokenAPIServer(c, fldPath)...)
+	return allErrs
+}
+
+func ValidateArgSelection(cfg *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(cfg.DiscoveryToken) != 0 && len(cfg.DiscoveryFile) != 0 {
+		allErrs = append(
+			allErrs,
+			field.Invalid(
+				fldPath,
+				nil,
+				"DiscoveryToken and DiscoveryFile cannot both be set",
+			),
+		)
 	}
-	if c.File != nil {
-		allErrs = append(allErrs, ValidateFileDiscovery(c.File, fldPath)...)
-		count++
+	if len(cfg.DiscoveryFile) == 0 {
+		if len(cfg.DiscoveryToken) == 0 {
+			cfg.DiscoveryToken = cfg.Token
+		}
+		if len(cfg.DiscoveryToken) == 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"Token value not found in DiscoveryToken or Token",
+				),
+			)
+		}
 	}
-	if c.HTTPS != nil {
-		allErrs = append(allErrs, ValidateHTTPSDiscovery(c.HTTPS, fldPath)...)
-		count++
+	if len(cfg.DiscoveryTokenAPIServers) < 1 {
+		allErrs = append(
+			allErrs,
+			field.Invalid(
+				fldPath,
+				nil,
+				"DiscoveryTokenAPIServers not set",
+			),
+		)
 	}
-	if count != 1 {
-		allErrs = append(allErrs, field.Invalid(fldPath, nil, "exactly one discovery strategy can be provided"))
+	// TODO remove once we support multiple api servers
+	if len(cfg.DiscoveryTokenAPIServers) > 1 {
+		fmt.Println("[WARN] we don't support multiple api servers yet")
 	}
 	return allErrs
 }
 
-func ValidateFileDiscovery(c *kubeadm.FileDiscovery, fldPath *field.Path) field.ErrorList {
+func ValidateTLSBootstrapTokenArg(cfg *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
+	if len(cfg.TLSBootstrapToken) != 0 {
+		if len(cfg.Token) != 0 {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					fldPath,
+					nil,
+					"TLSBootstrapToken or Token must be set",
+				),
+			)
+		}
+	}
 	return allErrs
 }
 
-func ValidateHTTPSDiscovery(c *kubeadm.HTTPSDiscovery, fldPath *field.Path) field.ErrorList {
+func ValidateJoinDiscoveryTokenAPIServer(c *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
+	for _, m := range c.DiscoveryTokenAPIServers {
+		_, _, err := net.SplitHostPort(m)
+		if err != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath, nil, err.Error()))
+		}
+	}
 	return allErrs
 }
 
-func ValidateTokenDiscovery(c *kubeadm.TokenDiscovery, fldPath *field.Path) field.ErrorList {
+func ValidateDiscoveryFile(c *kubeadm.NodeConfiguration, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
-	if len(c.ID) == 0 || len(c.Secret) == 0 {
-		allErrs = append(allErrs, field.Invalid(fldPath, nil, "token must be specific as <ID>:<Secret>"))
+	isURL := true
```

Comment:
ok, so here we have two cases:
1. `https://some-site.com/bar.conf`
2. `/etc/kubernetes/baz.conf`

I'd make the logic like this:
```go
u, err := url.Parse(c.DiscoveryFile)
if err != nil {
    // Ok, something without a URI scheme is passed, assume a file
    if _, err := os.Stat(c.DiscoveryFile); os.IsNotExist(err) {
         allErrs = append(allErrs, field.Invalid(fldPath, nil, "not a valid HTTPS URL or a file on disk"))
    }
} else {
    // Parsable URL, but require HTTPS
    if u.Scheme != "https" {
         allErrs = append(allErrs, field.Invalid(fldPath, nil, "must be https"))
    }
}
```

Then in defaulting, we might want to catch `file:///foo.conf` URLs as well, and then we should do something like this:
```go
// If the user passed a file URL, remove the file:// prefix
if len(obj.DiscoveryFile) != 0 {
    u, err := url.Parse(c.DiscoveryFile)
    if err == nil && u.Scheme == "file" {
        obj.DiscoveryFile = u.Path
    }
}
```

However, the latter suggestion is optional